borsh/src/schema_helpers.rs

```diff
@@ -36,3 +36,208 @@ pub fn schema_container_of<T: BorshSchema>() -> BorshSchemaContainer {
+        Ok(Definition::Struct { fields }) => match fields {
+            Fields::NamedFields(fields) => {
+                tuple(count, fields.into_iter().map(|(_, field)| field), defs)
+            }
+            Fields::UnnamedFields(fields) => tuple(count, fields, defs),
+            Fields::Empty => Some(0),
+        },
+
+        // Handle primitive types.  They have well-known sizes.
+        Err("nil") => Some(0),
+        Err("bool" | "i8" | "u8") => Some(count),
+        Err("i16" | "u16") => count.checked_mul(2),
+        Err("i32" | "u32" | "f32") => count.checked_mul(4),
+        Err("i64" | "u64" | "f64") => count.checked_mul(8),
+        Err("i128" | "u128") => count.checked_mul(16),
+
+        // Assume string with maximum length.  This is equivalent to `Vec<u8>`.
+        Err("string") => count.checked_mul(MAX_LEN.checked_add(4)?),
+
+        Err(_) => None,
+    }
+}
+
+#[test]
+fn test_max_serialized_size() {
+    #[cfg(not(feature = "std"))]
+    use alloc::string::String;
+
+    mod test_structs {
+        use crate as borsh;
+        #[cfg(not(feature = "std"))]
+        use alloc::string::ToString as _;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestEmpty;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestNamed {
+            _foo: usize,
+            _bar: [u8; 15],
+        }
+
+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestUnnamed(usize, [u8; 15]);
+    }
+
+    const MAX_LEN: usize = 4294967295;
+
+    assert_eq!(Some(2), max_serialized_size::<u16>());
+    assert_eq!(Some(8), max_serialized_size::<usize>());
+
+    assert_eq!(Some(1), max_serialized_size::<Option<()>>());
+    assert_eq!(Some(2), max_serialized_size::<Option<u8>>());
+    assert_eq!(
+        Some(9),
+        max_serialized_size::<core::result::Result<u8, usize>>()
+    );
+
+    assert_eq!(Some(0), max_serialized_size::<()>());
+    assert_eq!(Some(1), max_serialized_size::<(u8,)>());
+    assert_eq!(Some(5), max_serialized_size::<(u8, u32)>());
+
+    assert_eq!(Some(0), max_serialized_size::<[u8; 0]>());
+    assert_eq!(Some(16), max_serialized_size::<[u8; 16]>());
+    assert_eq!(Some(16), max_serialized_size::<[[u8; 4]; 4]>());
+
+    assert_eq!(Some(4 + MAX_LEN), max_serialized_size::<Vec<u8>>());
+    assert_eq!(Some(4 + MAX_LEN), max_serialized_size::<String>());
+
+    assert_eq!(None, max_serialized_size::<Vec<Vec<u8>>>());
+    assert_eq!(Some(4 + MAX_LEN * 4), max_serialized_size::<Vec<Vec<()>>>());
+    assert_eq!(
+        Some(0),
+        max_serialized_size::<[[[(); MAX_LEN]; MAX_LEN]; MAX_LEN]>()
+    );
+
+    use test_structs::*;
+
+    assert_eq!(Some(0), max_serialized_size::<TestEmpty>());
+    assert_eq!(Some(23), max_serialized_size::<TestNamed>());
+    assert_eq!(Some(23), max_serialized_size::<TestUnnamed>());
+    assert_eq!(None, max_serialized_size::<BorshSchemaContainer>());
+}
```

Comment:
@mina86 depending on what way to proceed is decided upon, a test for an outline of `BoundedVec<T, L, U>`, defined locally with an implementation of `BorshSchema`, may be added here too